### DISCUSSION TITLE
Crashing software for no reason

### DISCUSSION THREAD

**Discussion by Namba-ir** [miguelgrinberg/microdot#267] [Q&A]
> Hello my friend, I installed this very popular framework on nodemcu esp32s and designed a very beautiful site with it to manage my iot products with them, now the problem I have is:

The software is working, and when I want to go from one page to another, the software hangs for no reason.
Like the picture below:

![Screenshot 2024-11-18 163114](https://github.com/user-attachments/assets/a664523e-836b-4ad9-a1d4-2fb8ad7f3ea5)

My software settings are as follows:

`app = Microdot()
Response.default_content_type = 'text/html'
Response.default_send_file_max_age = 3600
freq(240000000)
`
I even tried to change the frequency of my CPU from 160 MHz to 240 MHz, so that maybe the problem is with the internal processing, but it didn't work.

After that, I thought maybe the problem was with the files, but I reduced my file size to below 5 kilobytes and used utemplate to maybe solve the problem, but it still didn't solve it.

Another solution:
Every time the function is executed, I tried to first empty the RAM space inside the microcontroller using the gc.collect() command, and I was still not satisfied with this, and I tried to use the time.sleep(3) command until the user It sends resets one after the other, the software or microcontroller does not crash, but the problem is still not solved!

Example Code :

`
# Medical
@app.route('/medical')
async def index(req):
    time.sleep()
    gc.collect()
    name = None
    return Template('Medical.html').render(name=name)
`

The last solution I tried was:
Instead of using app.run() directly, I used parallel processing in my code like below, but still didn't solve my problem.

`
async def main():
    await app.start_server(debug=True, port=80)
    
asyncio.run(main())
`

Finally, my problem is that the microdot version 2.0.8 software crashes for no reason, does not give an error, does not exit the software execution loop, or anything else.
I want to know exactly how I can solve my problem?!

Thank you very much in advance for giving me your precious time

**Comment by Namba-ir** [comment]
My dear friend, apart from the problem of crashing the software, I received another error related to RAM, which is now added to the previous problem:

Please tell me the final solution to solve these problems

![testsetsets](https://github.com/user-attachments/assets/57ca411e-d141-4e48-b90a-ebed9880279b)

**Reply by miguelgrinberg** [reply]
Please do not write issues on this project when you are asking questions about your own application. The issues board is only to report bugs. The discussions forum is the right place to ask questions. I have moved your question for you this time.

**Reply by Namba-ir** [reply]
Thank you very much, I apologize for not paying attention, can you please help me to solve my questions?!

**Comment by Namba-ir** [comment]
bro please help i am needed help.

**Comment by miguelgrinberg** [comment]
A `MemoryError` means that you are using all the memory that is available to the Python interpreter. The documentation has a section on what options can be used to reduce the memory usage of your application: https://microdot.readthedocs.io/en/latest/freezing.html.

**Comment by Namba-ir** [comment]
As an example, how should I use these codes?! It only explains how to install this library, but there is no sample code, please give me an example so I can learn how to use it, I really don't know how to use it.

**Reply by miguelgrinberg** [reply]
I'm sorry but I don't understand what example code do you need. This isn't about code, it is about optimizing memory usage in your device. Please read the documentation page I shared.

**Reply by Namba-ir** [reply]
I read the documentation page, but I did not understand at all, for example, how should I use this freezing, see, I have a series of sensors, and using your frame I am managing those sensors, the problem I have is that web pages sometimes Sometimes it doesn't open and the microdot web software stops working, this is my issue and it worries me a lot because I can't manage my sensors, now in the middle I want you to tell me how to use library freezing. I noticed that it comes exactly with the microdot library It freezes, but I'm using a whole other library and I really don't know what to do, I'm stuck.

**Reply by miguelgrinberg** [reply]
Freezing is a micropython topic, it is not related to microdot. You need to look at the documentation for your device for building instructions.

**Reply by Namba-ir** [reply]
ok thank you very much
